### PR TITLE
The war of symbol and symbol (the PR dedicated to duplicate symbols breaking rustc in unexpected ways)

### PR DESCRIPTION
**Pull request by nagisa**
We provide tools to tell what exact symbols to emit for any fn or static, but
don’t quite check if that won’t cause any issues later on. Some of the issues
include LLVM mangling our names again and our names pointing to wrong locations,
us generating dumb foreign call wrappers, linker errors, extern functions
resolving to different symbols altogether (`extern {fn fail();} fail();` in some
cases calling `fail1()`), etc.

Before the commit we had a function called `note_unique_llvm_symbol`, so it is
clear somebody was aware of the issue at some point, but the function was barely
used, mostly in irrelevant locations.

Along with working on it I took liberty to start refactoring trans/base into
a few smaller modules. The refactoring is incomplete and I hope I will find some
motivation to carry on with it.

This is possibly a [breaking-change] because it makes dumbly written code
properly invalid.

This fixes all those issues about incorrect use of #[no_mangle] being not reported/misreported/ICEd by the compiler.

NB. This PR does not attempt to tackle the parallel codegen issue that was mentioned in https://github.com/rust-lang/rust/pull/22811, but I believe it should be very straightforward in a follow up PR by modifying `trans::declare::get_defined_value` to look at all the contexts.

cc @alexcrichton @huonw @nrc because you commented on the original RFC issue.

EDIT: wow, this became much bigger than I initially intended.

### CODE REVIEW

File: src/librustc_trans/trans/base.rs [hunk 1]
```diff
@@ -2253,27 +2009,24 @@ pub fn trans_item(ccx: &CrateContext, item: &ast::Item) {
...
-                             &item.attrs);
+                    trans_fn(ccx, &**decl, &**body, llfn, empty_substs, item.id, &item.attrs);
                 }
-                update_linkage(ccx,
-                               llfn,
-                               Some(item.id),
+                update_linkage(ccx, llfn, Some(item.id),
                                if is_origin { OriginalTranslation } else { InlinedCopy });
+
+                if is_entry_fn(ccx.sess(), item.id) {
```
Comment (nagisa) [inline src/librustc_trans/trans/base.rs]: NB: before the move of this block from `finish_register_fn` we reported “compilation successful” error before we created the entry wrapper or actually translated contents of the entry function.

File: src/librustc_trans/trans/base.rs [hunk 2]
```diff
@@ -2608,27 +2179,36 @@ pub fn is_entry_fn(sess: &Session, node_id: ast::NodeId) -> bool {
...
+/// Create the `main` function which will initialise the rust runtime and call users’ main
+/// function.
 pub fn create_entry_wrapper(ccx: &CrateContext,
                            _sp: Span,
                            main_llfn: ValueRef) {
     let et = ccx.sess().entry_type.get().unwrap();
     match et {
         config::EntryMain => {
-            create_entry_fn(ccx, main_llfn, true);
+            create_entry_fn(ccx, _sp, main_llfn, true);
```
Comment (alexcrichton) [inline src/librustc_trans/trans/base.rs]: Could you rename this variable to `sp` to show that it's used now?

File: src/librustc_trans/trans/base.rs [hunk 3]
```diff
@@ -2608,27 +2179,36 @@ pub fn is_entry_fn(sess: &Session, node_id: ast::NodeId) -> bool {
...
         }
-        config::EntryStart => create_entry_fn(ccx, main_llfn, false),
+        config::EntryStart => create_entry_fn(ccx, _sp, main_llfn, false),
         config::EntryNone => {}    // Do nothing.
     }
 
+    #[inline(never)]
```
Comment (alexcrichton) [inline src/librustc_trans/trans/base.rs]: Is this still needed?

File: src/librustc_trans/trans/base.rs [hunk 4]
```diff
@@ -2745,15 +2324,14 @@ pub fn get_item_val(ccx: &CrateContext, id: ast::NodeId) -> ValueRef {
 
             let v = match i.node {
                 ast::ItemStatic(_, _, ref expr) => {
-                    // If this static came from an external crate, then
-                    // we need to get the symbol from csearch instead of
-                    // using the current crate's name/version
-                    // information in the hash of the symbol
+                    // If this static came from an external crate, then we need to get the symbol
+                    // from csearch instead of using the current crate's name/version information
+                    // in the hash of the symbol
```
Comment (alexcrichton) [inline src/librustc_trans/trans/base.rs]: It'll help reduce the diff if comments like this are not reformatted. I also personally dislike widening comments. Our style guidelines say that code can take up to 100 characters, not that it needs to take up to 100 characters. If no other changes are made I believe it to be best to leave it as-is.

File: src/librustc_trans/trans/declare.rs [hunk 5]
```diff
@@ -0,0 +1,263 @@
...
+//! directly. These functions do some additional work to ensure we do the right thing given
+//! the preconceptions of trans.
+//!
+//! Some useful guidelines:
+//!
+//! * Use declare_* family of methods if you are declaring, but are not interested in defining the
+//! ValueRef they return.
+//! * Use define_* family of methods when you might be defining the ValueRef.
+//! * When in doubt, define.
+#![allow(dead_code)]
```
Comment (alexcrichton) [inline src/librustc_trans/trans/declare.rs]: Is this temporary?

**Comment by alexcrichton** [comment]
This seems like a fine refactoring to me, thanks @nagisa! I'm not trans hacker though so I'm not comfortable r+'ing

**Comment by nagisa** [comment]
Thanks for taking a look @alexcrichton!

File: src/librustc_trans/trans/attributes.rs [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+                }
+                _ => {}
+            }
+
+            // We can also mark the return value as `dereferenceable` in certain cases
+            match ret_ty.sty {
+                // These are not really pointers but pairs, (pointer, len)
+                ty::ty_uniq(it) |
+                ty::ty_rptr(_, ty::mt { ty: it, .. }) if !common::type_is_sized(ccx.tcx(), it) => {}
+                ty::ty_uniq(inner) | ty::ty_rptr(_, ty::mt { ty: inner, .. }) => {
```
Comment (arielb1) [inline src/librustc_trans/trans/attributes.rs]: Why the double negative?

Comment (nagisa) [inline src/librustc_trans/trans/attributes.rs]: This function has only been moved and renamed from `base::get_fn_llvm_attributes` only. The body has seen no changes (other than variable renames, of course).

Nicely spotted, nevertheless. I will go ahead and fix this.

**Comment by pnkfelix** [comment]
(i'm going to be out until wedneday; but I will try to get to this then.  sorry for the absurd delay; maybe you can rebase in the meantime @nagisa )

**Comment by nagisa** [comment]
Rebased. I didn’t even notice the last breakage until now, didn’t get github mail.

**Comment by nagisa** [comment]
Rebased.

**Comment by pnkfelix** [comment]
@nagisa my plan is to review this right after the beta-release, is that okay with you?

**Comment by nagisa** [comment]
Sure, been waiting for a month, can easily wait a few weeks more. I’ll have to rigorously review the patch for bitrot, though.

**Comment by pnkfelix** [comment]
@nagisa yeah, sorry about the delay...

**Pull request merged**